tests/test_database.py

```diff
@@ -0,0 +1,488 @@
+    r = dict()
+    if entrypoint == "db":
+        commit_to_session(session, table)
+        # Need to `get` b/c db doesn't return a response
+        model_db = session.get(models.table, table.id)
+        data = model_db.dict()
+        r.update({entrypoint: data})
+    elif entrypoint == "abstract-funcs":
+        model_db = abstractions.create(
+            session=session, table_cls=models.table, model=models.table(**request)
+        )
+        data = model_db.dict()
+        r.update({entrypoint: data})
+    elif entrypoint == "client":
+        client = Client(base_url=http_server)
+        response = client.post(models.path, json=request)
+        assert response.status_code == 200
+        data = response.json()
+        r.update({entrypoint: data})
+    elif entrypoint == "cli":
+        data = get_data_from_cli("post", http_server, models.path, request)
+        r.update({entrypoint: data})
+
+    # evaluate data
+    for k in request.keys():
+        if isinstance(r[entrypoint][k], datetime):
+            assert r[entrypoint][k] == parse_to_datetime(request[k])
+        elif (
+            # Pydantic requires a "Z"-terminated timestamp, but FastAPI responds without the "Z"
+            isinstance(r[entrypoint][k], str)
+            and isinstance(request[k], str)
+            and any([s.endswith("Z") for s in (r[entrypoint][k], request[k])])
+            and not all([s.endswith("Z") for s in (r[entrypoint][k], request[k])])
+        ):
+            assert add_z(r[entrypoint][k]) == add_z(request[k])
+        else:
+            assert r[entrypoint][k] == request[k]
+    if blank_opts:
+        for k in blank_opts:
+            assert r[entrypoint][k] is None
+    assert r[entrypoint]["id"] is not None
+
+
+@pytest.mark.parametrize("entrypoint", ENTRYPOINTS)
+def test_create_incomplete(session, model_to_create, entrypoint, http_server):
+    models, request, _ = model_to_create
+    # make sure the database is empty
+    clear_table(session, models.table)
+
+    incomplete_request = copy.deepcopy(request)
+    # Remove a required field
+    del incomplete_request[list(incomplete_request)[0]]
+    table = models.table(**incomplete_request)
+
+    if entrypoint == "db":
+        with pytest.raises(IntegrityError):
+            commit_to_session(session, table)
+    elif entrypoint == "abstract-funcs":
+        with pytest.raises(ValidationError):
+            _ = abstractions.create(session=session, table_cls=models.table, model=table)
+    elif entrypoint == "client":
+        client = Client(base_url=http_server)
+        response = client.post(models.path, json=incomplete_request)
+        assert response.status_code == 422
+    elif entrypoint == "cli":
+        data = get_data_from_cli("post", http_server, models.path, incomplete_request)
+        error = data["detail"][0]
+        assert error["msg"] == "field required"
+        assert error["type"] == "value_error.missing"
```

Comment:
Perhaps the solution is to have `entrypoint` not be a string but rather a parametrized fixture that returns a function, similar to `execute_recipe` in pangeo-forge-recipes. Then you could just do something like

```python
entrypoint(table)
```

Reply:
☑️ [`CreateFixtures`](https://github.com/cisaacstern/pangeo-forge-orchestrator/blob/5141edce2557550cd12ba8da07b03d6b73218f1b/tests/conftest.py#L176-L209) defines a parametrized fixture (`create_func`) which returns a function, which is then passed via class inheritance to [`test_create`](https://github.com/cisaacstern/pangeo-forge-orchestrator/blob/5141edce2557550cd12ba8da07b03d6b73218f1b/tests/test_database.py#L142). The same pattern is now followed for each of the CRUD test groups:

1. `<Operation>Fixtures` class defines parametrized fixture which returns a function
2. `Test<Operation>` class inherits from fixture class